File: src/so/aporia/utils/user/render/animation/SpringSimulator.java
package so.aporia.utils.user.render.animation;

/**
 * Real-time spring physics simulator.
 * <p>
 * Models a damped harmonic oscillator: {@code F = -k*x - d*v}.
 * Call {@link #update(float)} every frame with delta-time in seconds,
 * then read {@link #value()} and {@link #velocity()}.
 * <p>
 * Usage:
 * <pre>{@code
 * SpringSimulator spring = new SpringSimulator(200f, 20f, 0f);
 * spring.setTarget(1f);
 * // each frame:
 * spring.update(deltaSeconds);
 * float x = spring.value();
 * }</pre>
 */
public final class SpringSimulator {

    private float stiffness; /* k — spring constant */
    private float damping;   /* d — damping coefficient */
    private float mass;      /* m — mass (default 1) */

    private float position;
    private float velocity;
    private float target;

    /**
     * @param stiffness spring constant k (try 150–400 for snappy UI)
     * @param damping   damping d (try 15–30; critical = 2*sqrt(k*m))
     * @param initial   starting position
     */
    public SpringSimulator(float stiffness, float damping, float initial) {
        this.stiffness = stiffness;
        this.damping   = damping;
        this.mass      = 1f;
        this.position  = initial;
        this.target    = initial;
    }

    /** Advances the simulation by {@code dt} seconds. */
    public void update(float dt) {
        if(dt <= 0) return;
        /* Semi-implicit Euler integration — stable for large dt. */
        float force = -stiffness * (position - target) - damping * velocity;
        velocity += (force / mass) * dt;
        position += velocity * dt;
    }

    /** Convenience: update with milliseconds. */
    public void updateMs(long ms) { update(ms / 1000f); }

    public float value()    { return position; }
    public float velocity() { return velocity; }

    public void setTarget(float t)   { this.target   = t; }
    public void setPosition(float p) { this.position = p; this.velocity = 0; }
    public void snap(float p)        { this.position = p; this.target = p; this.velocity = 0; }

    public float getTarget()    { return target; }
    public boolean isSettled()  { return Math.abs(position - target) < 0.001f && Math.abs(velocity) < 0.001f; }

    /* --- Presets --- */

    /** Snappy UI spring — fast, slight overshoot. */
    public static SpringSimulator snappy(float initial) {
        return new SpringSimulator(300f, 22f, initial);
    }

    /** Bouncy spring — noticeable oscillation. */
    public static SpringSimulator bouncy(float initial) {
        return new SpringSimulator(200f, 10f, initial);
    }

    /** Smooth spring — no overshoot, gentle. */
    public static SpringSimulator smooth(float initial) {
        return new SpringSimulator(120f, 24f, initial);
    }

    /** Stiff spring — almost instant, tiny overshoot. */
    public static SpringSimulator stiff(float initial) {
        return new SpringSimulator(500f, 35f, initial);
    }

    /** Wobbly spring — lots of oscillation. */
    public static SpringSimulator wobbly(float initial) {
        return new SpringSimulator(180f, 8f, initial);
    }
}

File: src/so/aporia/utils/user/render/animation/SpringSimulator.kt
/*
 * Copyright (c) 2025-2026 Aporia.cc Project
 * Distributed under the Aporia.cc Software License Agreement v1.0
 * See LICENSE and COPYRIGHT files in the project root for full text.
 */

package so.aporia.utils.user.render.animation

import kotlin.math.abs
import com.chaos.annotation.ChaosNative
/**
 * Real-time spring physics simulator.
 *
 * Models a damped harmonic oscillator: `F = -k*x - d*v`.
 * Call [update] every frame with delta-time in seconds,
 * then read [value] and [velocity].
 *
 * Usage:
 * ```kotlin
 * val spring = SpringSimulator(200f, 20f, 0f)
 * spring.setTarget(1f)
 * // each frame:
 * spring.update(deltaSeconds)
 * val x = spring.value
 * ```
 */
@ChaosNative
class SpringSimulator(
    private var stiffness: Float,
    private var damping: Float,
    initial: Float
) {

    private var mass = 1f
    private var position = initial
    private var velocity = 0f
    private var target = initial

    fun update(dt: Float) {
        if (dt <= 0) return
        val force = -stiffness * (position - target) - damping * velocity
        velocity += (force / mass) * dt
        position += velocity * dt
    }

    fun updateMs(ms: Long) { update(ms / 1000f) }

    fun value(): Float = position
    fun velocity(): Float = velocity

    fun setTarget(t: Float) { target = t }
    fun setPosition(p: Float) { position = p; velocity = 0f }
    fun snap(p: Float) { position = p; target = p; velocity = 0f }

    fun getTarget(): Float = target
    fun isSettled(): Boolean = abs(position - target) < 0.001f && abs(velocity) < 0.001f

    companion object {
        @JvmStatic fun snappy(initial: Float): SpringSimulator = SpringSimulator(300f, 22f, initial)
        @JvmStatic fun bouncy(initial: Float): SpringSimulator = SpringSimulator(200f, 10f, initial)
        @JvmStatic fun smooth(initial: Float): SpringSimulator = SpringSimulator(120f, 24f, initial)
        @JvmStatic fun stiff(initial: Float): SpringSimulator = SpringSimulator(500f, 35f, initial)
        @JvmStatic fun wobbly(initial: Float): SpringSimulator = SpringSimulator(180f, 8f, initial)
    }
}
